src/log-store/src/raft_engine/log_store.rs

```diff
@@ -0,0 +1,545 @@
+            recovery_mode: RecoveryMode::TolerateTailCorruption,
+            batch_compression_threshold: ReadableSize::kb(8),
+            target_file_size: ReadableSize(config.max_log_file_size as u64),
+            ..Default::default()
+        };
+        let engine = Arc::new(Engine::open(raft_engine_config).context(RaftEngineSnafu)?);
+        Ok(Self {
+            config,
+            engine,
+            cancel_token: Mutex::new(None),
+            gc_task_handle: Mutex::new(None),
+            started: AtomicBool::new(false),
+        })
+    }
+
+    pub fn started(&self) -> bool {
+        self.started.load(Ordering::Relaxed)
+    }
+}
+
+impl Debug for RaftEngineLogstore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RaftEngineLogstore")
+            .field("config", &self.config)
+            .field("started", &self.started.load(Ordering::Relaxed))
+            .finish()
+    }
+}
+
+#[async_trait::async_trait]
+impl LogStore for RaftEngineLogstore {
+    type Error = Error;
+    type Namespace = Namespace;
+    type Entry = Entry;
+
+    async fn start(&self) -> Result<(), Self::Error> {
+        let engine_clone = self.engine.clone();
+        let interval = self.config.gc_interval;
+        let token = CancellationToken::new();
+        let child = token.child_token();
+        // TODO(hl): Maybe spawn to a blocking runtime.
+        let handle = common_runtime::spawn_bg(async move {
+            loop {
+                tokio::select! {
+                    _ = tokio::time::sleep(interval) => {}
+                    _ = child.cancelled() => {
+                        info!("LogStore gc task has been cancelled");
+                        return;
+                    }
+                }
+                match engine_clone.purge_expired_files().context(RaftEngineSnafu) {
+                    Ok(res) => {
+                        // TODO(hl): the retval of purge_expired_files indicates the namespaces need to be compact,
+                        // which is useful when monitoring regions failed to flush it's memtable to SSTs.
+                        info!(
+                            "Successfully purged logstore files, namespaces need compaction: {:?}",
+                            res
+                        );
+                    }
+                    Err(e) => {
+                        error!(e; "Failed to purge files in logstore");
+                    }
+                }
+            }
+        });
+        *self.cancel_token.lock().await = Some(token);
+        *self.gc_task_handle.lock().await = Some(handle);
+        self.started.store(true, Ordering::Relaxed);
+        info!("RaftEngineLogstore started with config: {:?}", self.config);
+        Ok(())
+    }
+
+    async fn stop(&self) -> Result<(), Self::Error> {
+        ensure!(self.started(), IllegalStateSnafu);
+        self.started.store(false, Ordering::Relaxed);
```

Comment:
The above two lines of code are not atomic operations, so the following code may still be called multiple times.

Reply:
Should use `compare_and_exchange` instead.